DataFormats/Provenance/src/IndexIntoFile.cc

```diff
@@ -1980,27 +2030,408 @@ namespace edm {
+             info.eventSequenceRunOrLumiEntry_->orderPHIDRun() == currentRun) {
+        addToFileOrder(info.eventSequenceIndex_, false, true);
+        info.nextEventSequence(runOrLumiEntries);
+      }
+    }
+  }
+
+  void IndexIntoFile::IndexIntoFileItrEntryOrder::handleToEndOfContiguousEventsInLumis(
+      EntryOrderInitializationInfo& info, EntryNumber_t currentRun, int endOfRunEntries) {
+    // Form a list of lumis that have no more events left to be processed and are in the current
+    // run and have at least one valid Lumi TTree entry. Contains the index to the first
+    // lumi entry and its TTree entry number, sorted by earliest lumi TTree entry number.
+    std::vector<TTreeEntryAndIndex> lumisWithNoRemainingEvents;
+    fillLumisWithNoRemainingEvents(
+        lumisWithNoRemainingEvents, endOfRunEntries, currentRun, info.eventSequenceRunOrLumiEntry_);
+    auto nextLumiWithNoEvents = lumisWithNoRemainingEvents.cbegin();
+    auto endLumisWithNoEvents = lumisWithNoRemainingEvents.cend();
+
+    // On each step of this iteration we process all the events in a contiguous sequence of events
+    // from a single lumi (these are events that haven't already been processed and are contained
+    // within the last contiguous sequence of events from the containing run).
+    while (info.iEventSequence_ < info.iEventSequenceEnd_ &&
+           info.eventSequenceRunOrLumiEntry_->orderPHIDRun() == currentRun) {
+      auto currentLumi = info.eventSequenceRunOrLumiEntry_->orderPHIDRunLumi();
+
+      // Last contiguous sequence of events in lumi
+      if (info.startOfLastContiguousEventsInLumi_[currentLumi] == info.eventSequenceIndex_) {
+        auto firstBeginEventsContiguousLumi = info.eventSequenceRunOrLumiEntry_->beginEvents();
+        // Find the first Lumi TTree entry number for this Lumi
+        EntryNumber_t lumiTTreeEntryNumber = invalidEntry;
+        setToLowestInLumi(lumiTTreeEntryNumber, info, currentLumi);
+
+        // In addition, we want lumis before this in the lumi tree if they have no events
+        // left to be processed
+        handleLumisWithNoEvents(nextLumiWithNoEvents, endLumisWithNoEvents, lumiTTreeEntryNumber);
+
+        // Handle the lumi with the next sequence of events to process
+        handleLumiWithEvents(info, currentLumi, firstBeginEventsContiguousLumi);
+
+      } else {
+        // not last contiguous event sequence for lumi
+        while (info.iEventSequence_ < info.iEventSequenceEnd_ &&
+               info.eventSequenceRunOrLumiEntry_->orderPHIDRunLumi() == currentLumi) {
+          addToFileOrder(info.eventSequenceIndex_, false, true);
+          info.nextEventSequence(indexIntoFile()->runOrLumiEntries());
+        }
+      }
+    }
+    handleLumisWithNoEvents(nextLumiWithNoEvents, endLumisWithNoEvents, invalidEntry, true);
+  }
+
+  void IndexIntoFile::IndexIntoFileItrEntryOrder::setToLowestInLumi(EntryNumber_t& lumiTTreeEntryNumber,
```

Comment:
This should be `const` and it should return the `lumiTTreeEntryNumber` instead of being passed one.

Reply:
Done. That is better. Thanks.